man/man1/fsviewer.1
.\" Man page for fsviewer
.\"
.\" Hacked from the:
.\" Man page for fsviewer
.\" Copyright (C), 1994, 1995, Graeme W. Wilford. (Wilf.)
.\"
.\" You may distribute under the terms of the GNU General Public
.\" License 
.\"
.TH fsviewer 1x "July 1st, 1999" "NeXT fileviewer lookalike"
.SH NAME
fsviewer \- a NeXT fileviewer lookalike
.SH SYNOPSIS
.\" The general command line
.B fsviewer 
.SH DESCRIPTION
.B fsviewer 
is a NeXT fileviewer lookalike designed to work with
the Window Maker window manager. It features 
Browser mode, DND and file operations.

.SH "THE INTERFACE"

The interface is split into 3 sections. 
The shelf is an area to place frequently accessed directories 
or files. The pathview section depicts the current selected path and the 
browser columns lists the directory contents of the current selected path. 
Icons/Actions/Events can be set in the Inspector Panel.

.SH "KEY BINDINGS"
All keys bindings can only be accessed in combination with the Meta key.
In systems without a Meta key, this should be the Alt key.

.SH "MOUSE CLICKS"
Clicking on an entry in a browser column will update the scroll view with the
new selection and if the selection is a directory, it will fill the next 
column. It will then scroll the browser/pathview accordingly.

Double-clicking on an entry in the pathview/browser will execute it if it is 
executable or launch the associated editor/viewer. 

Ctrl-MouseClick over an icon in the shelf will remove it from the shelf.

.SH INSPECTOR
The inspector contains 6 different panels:

.Bl -bullet
.It Attributes
The attributes panel displays the file permissions, the last changed date,
file size, owner and group. The link section is not working.

.Bl -bullet
.It Icon
The icon panel allows an icon to be selected for the current extension.  If 
the file does not have an extension it will be associated with the full
filename. The paths used are read from the Window Maker defaults domain. To add
paths either edit the Window Maker defaults file or use WPrefs.app. 
Click 'OK' to set the new icon. 

.Bl -bullet
.It Viewer
The viewer panel allows for the selection of the default viewer to be 
associated with the file extension. The list all the executables registered 
with FSViewer.

.Bl -bullet
Editor
4) The editor is the same as the Viewer. This becomes useful when a
file requires a separate editor e.g. HTML files.

.Bl -bullet 
.It Executable
The executable panel allows a command to be associated with the file. 
This panel is only accessible when the file is executable. Three shortcuts 
have been provided. %p for the current path, %f for the current filename and 
%s for the filename to be passed as an arg to the program. For example if 
/usr/bin/xemacs was selected:

       %p = /usr/bin/
       %f = xemacs

And a possible command could be:

       %p%f %s

which translates to

       /usr/bin/xemacs %s

Any file that has xemacs as it's default viewer/editor will use this command
where %s is replaced with the full path and filename.

.Bl -bullet
.It File Extensions
The file extension panel provides a method for associating file extensions
with an app. An extension MUST start with a '.'. If a new extension 
is entered its icon may be set from this panel.

.SH "Magic Files"
Rudimentary magic file support is available -- the app will try to 
figure out what type the file is if it is unknown. It will then do something
based on the file type. There are 3 cases:

.TP
If it is ascii, xedit is started.
.TP
If it is an image, xv is started.
.TP
If it is a ps file, ghostview is started.

To change these settings, edit the defaults file in 
.I ~/GNUstep/Defaults.


.SH DnD
A port of the OffiX DnD library is being used in this app. The port has been 
undertaken by Igor Roboul. OffiX DnD is supported by Window Maker so you can
drop things on the dock/clip. Also icons can be dragged from the pathview to a 
folder or to the shelf, dragging to a folder causes the item to be moved into
that folder. Presing the Shift key while dragging will cause the item to be
copied.


.SH "Mounting Devices"

The app provides rudimentary support for mounting, unmounting and ejecting 
file systems. A sample defaults file entry is provided below.

  DISKS = {
    devices = ((CDROM, "/cdrom", "/dev/hdc"), \\
    (Zip, "/dosz", "/dev/sda4"));
    mount = "mount %s";
    umount = "umount %s";
    eject = "eject %s";
  };

At the moment any errors from the commands are not detected.


.SH "File Settings"
When an unknown file type is encountered a window will pop up prompting you
for a command. For example if the file is a jpeg the command may be:

       xv %s

where %s will be replaced with the filename when the app comes across a jpeg 
type file. To pass something as one arguement, just wrap it in double quotes
eg:

       app "-p foo"

The "-p foo" will be passed as one arguement.


.SH Preferences
Sometimes it may be easier to edit the defaults file 
(
.I ~/GNUstep/Defaults/FSViewer
) than to work with the app. Also there are some 
options that the currently sparse Preferences Dialog does not know about. 
It is worth looking at the file to see what can be changed. A more 
comprehensive Preferences Dialog Box, that will allow access to most if not all
of the defaults file, is planned for the next version.
Make sure to backup the defaults file if you plan to edit it by hand!

.SH FILES
.TP
.I /etc/GNUstep/Defaults/FSViewer
the systemwide configuration file.
.TP
.I ~/GNUstep/Defaults/FSViewer
a user's personal configuration file.

.SH BUGS
The link section under Attributes doesn't work.

Many features remain unimplemented.  A list of these from the 
.B fsviewer
README file:
.TP
Revert in the Inspector Panel (partially supported).
.TP 
Recycler Support.
.TP
Implement a locate/find dialog. Done but it needs to be customizable.
.TP
Improve the feel of the app.
.TP
Add a panel to allow for 'quick' registering of apps.
.TP
Change cursor during busy periods.
.TP
Add in support for file completion for navigating the columns.
.TP
Provide an interface to PGP to allow for file encryption via FSViewer.
.TP
Implement other viewing modes.
.TP
Implement a full preferences dialog.
.TP
Revert in the Inspector Panel (partially supported).

Send comments and bug reports to <clernong@tinet.ie>.

.SH "SEE ALSO"
.BR wmaker (1x), 
.SH AUTHORS
.B fsviewer 
is Copyright 1998/1999 George Clernon <clernong@tinet.ie>
and was written by George Clernon.

This manpage was adapted from the 
.B fsviewer
README file by Will Lowe <lowe@debian.org> for the Debian GNU/Linux distribution.  


ChangeLog
## NEXT

- Fix secondary windows to being resizable. ([#15](https://github.com/roblillack/fsviewer/pull/15))
- Fix showing correct free disk space with SI prefixes. ([#17](https://github.com/roblillack/fsviewer/pull/17))
- Fix alpha channel blending for icons everywhere. ([#18](https://github.com/roblillack/fsviewer/pull/18))
- Fix crash when when directory contains Umlaut entries on macOS ([#21](https://github.com/roblillack/fsviewer/pull/21))
- Add support for PNG icons ([#19](https://github.com/roblillack/fsviewer/pull/19))

## FSViewer 0.2.7 (2024-06-22)

- Switch build system to cmake
- Fix build to work with modern versions of WINGs & Window Maker
- Adapt window bottom to more closely resemble NeXT FileViewer
- Add support for showing available space on filesystem
- Fix recurring crashes
- Fix main window not showing title bar
- Fix building on 64 bit systems
- Fix building on FreeBSD
- Fix building on OpenBSD

## FSViewer 0.2.6 (2007-10-14)
Latest version done by Guido Scholz <guido.scholz@bayernline.de>

	* po/sv.po: Swedish translation added.
	* man/fsviewer.x1, chdef.x1: Man pages from debian package added.
	* configure.sub, configure.guess: updated autoconf files
	* configure.ac: some administrative changes, error messages for
	not found libraries etc.
	* chdef/chdef.c: Removed dependency from "proplist-compat.h".
	* FSPrefs.c: Update file view if view options are changed
	activating the apply button.

FSViewer.app-0.2.5 (20021215)
-----------------------------
	o All lists in preferencs panel and inspector panel are sorted
	  in alphabetical order.
	o Added a russian (ru) translation.
	o Changed gettext DOMAIN from "FSViewer.app" to "FSViewer".
	o Removed dependency from "proplist-compat.h".
	o Comment out <sys/rlist.h> in FSUtils.c for FreeBSD 4.6.
	o Revised the german (de) translation (now more "NeXTish").
	o Changes in attribute inspector panel:
	  - Link label now changes color (black/darkgrey)
	  - File size is displayed in Bytes/KB/MB
	o A selection change in file viewer lists now updates the
	  inspector panel.


FSViewer.app-0.2.4 (20021110)
-----------------------------
	o Some administrative changes to remove the dist-bzip2 target
	  due to my configuration mess with autotools.
	o Removed double entries for file extensions in prop.c and
	  sorted them in alphabetical order.
	o Implementation of i18n support (gettext).
	o Added a german (de) translation.
	o Much work on the inspector panel, e.g.:
	  - link targets are shown
	  - long label texts are (mostly) wraped
	  - some rearrangements to make label dimensions
	    fit (dynamically) to translated texts
	  - undo for changes to file attributes is implemented
	o Text in legal panel is wraped.
	o Reintroduced (lost) images for home and neighbour.
	o Icons in panels are now painted with
	  WMCreateBlendedPixmapFromFile, which removes the dithering
	  shade around icons and solves errors with the home icon.


FSViewer.app-0.2.3f (20021025)
-----------------------------
	o Crashfix in FSPrefs.c, now respecting NULL pointers to
	  property list dictionary entries.


FSViewer.app-0.2.3e (20021009)
-----------------------------
	o Changes in FSViewer.c to handle wrong values of XPOS, YPOS,
	  WSIZE and HSIZE without crash when read from property list.
	o Some administrative changes to make the FSViewer package
	  distributable as bzip2 file.
	o Substitution of some broken pixmaps.


FSViewer.app-0.2.3d (20020225)
-----------------------------
	o Removed FSVERSION in FSViewer.h, changed FSPanel.c,
	  FSViewer.c and configure.in to use VERSION defined in
	  configure.in.
	o Integrated xpm and tiff directories with containing icons.
	o Fixed bug in prop.c which crashes FSViewer writing the user
	  default database.


FSViewer.app-0.2.3c (20020107)
-----------------------------
	o Fixed bug in FSUtils.c (was complained by gcc 2.96.0).


FSViewer.app-0.2.3b (20020105)
-----------------------------
	o Made adaptation to WindowMaker 0.80.0.
	o Removed libPropList dependency.


FSViewer.app-0.2.3 (19991010)
-----------------------------
	o Changed MAXINT to INT_MAX for NetBSD v1.3.2
	o Added Preferences Window
	o Fixed bug in FSGetStringForName, if the returned
	  value is NULL, wstrdup is not called.
	o Fixed bug in FSCreatePixmapWithBacking*** when 
	  image is bigger than backing.
	o Created Icon Selection Dialog.
	o Updated Icon Inspector to reflect look of Icon Selection
	  Dialog.
	o Set window level for some dialogues.
	o Updated Edit|View Inspectors to save the extn into the EXTN
	  defaults file entry.
	o Modified chdefs to make the latest changes the necessary
	  changes to the defaults file.
	o Added DisplayMCListPixmap variable to turn on/off display
	  of pixmaps indicating file/directory. The list runs better with
	  it off!
	o Add check for empty clientData in paintItem of FSMCList.
	
	
FSViewer.app-0.2.2 (19990731)
-----------------------------
	o Added icons for files and directories in list view
	o Plugged a few memory leaks found by Electric Fence
	o Insertion/Deletion of files now updates all views
	o Changed the configure script to use get-wraster-flags,
	  this will hopefully fix most "can't find library" problems
	o Changed listview parsing method.
	o Applied a patch by Igor Roboul to FSViewer.c and iconInspector.c 
	
	
FSViewer.app-0.2.1 (19990622)
-----------------------------
	o FSViewer now compiles with WM 0.60
	o Added a preliminary list view. This is really for v0.3.0 but I left
	  it in anyway.
	o FSGetStringFor*** now returns a duplicated string.
	o wmallocing memory with (Disk) instead of (Disk *)
	o Fixed icon naming problem when the icon changes but the popup
	  doesn't
	
	
FSViewer.app-0.2.0 (19990519)
-----------------------------
	o Added dialog to prompt for settings with respect to unknown
	  file types.
	o Added app specific alert panel.
	o Added in fallback condition that if the creation of an icon fails
	  it defaults to the file_plain xpm.

src/FSViewer.c
/* FSViewer.c  */

#include <WINGs/WINGs.h>
#include <WMaker.h>
#include <X11/Xlib.h>
#include <X11/Xlocale.h>
#include <X11/Xproto.h>
#include <X11/Xresource.h>
#include <X11/Xutil.h>
#include <X11/keysym.h>
#include <locale.h>
#include <unistd.h> /* for access() */

#include "FSFileButton.h"
#include "FSFileView.h"
#include "FSFinder.h"
#include "FSMenu.h"
#include "FSUtils.h"
#include "FSViewer.h"
#include "filebrowser.h"
#include "windowmanager.h"
#include "xpm/FSViewer.xpm"

#define DEBUG 0

/* Application defaults */
WMUserDefaults* defaultsDB;
WMPropList* filesDB;

static FSViewer* fsViewer;
static Bool focusIn;
int ModifierFromKey(Display* dpy, char* key);

void wAbort(Bool foo)
{
    exit(1);
}

static void
FSLoadFSViewerConfigurations(FSViewer* fsViewer)
{
    char* iconDir = NULL;
    char* magicFile = NULL;

    defaultsDB = WMGetStandardUserDefaults();

    /*if there is no user database available, create one*/
    if (WMGetUDStringForKey(defaultsDB, "DIRECTORY") == NULL) {
        InitFilesDB(fsViewer);
    }
    iconDir = FSGetStringForName("ICONDIR");

    if (iconDir) {
        /* test for existing ICONDIR */
        if (access(iconDir, F_OK) != 0) {
            wwarning(_("User database entry for ICONDIR is not "
                       "valid: %s. Taking default value."),
                iconDir);
            free(iconDir);

            /*try compile time default value*/
            WMSetUDStringForKey(defaultsDB, ICONDIR, "ICONDIR");
            iconDir = FSGetStringForName("ICONDIR");

            if (iconDir) {
                if (access(iconDir, F_OK) != 0) {
                    wwarning(_("Default ICONDIR not found: %s"), iconDir);
                    free(iconDir);
                    exit(1);
                } else {
                    WMSetResourcePath(iconDir);
                    free(iconDir);
                }
            } else {
                wwarning(_("No valid value for ICONDIR found. "
                           "Even default directory is broken!"));
                exit(1);
            }

        } else {
            WMSetResourcePath(iconDir);
            free(iconDir);
        }
    }

    else {
        wwarning(_("No value for ICONDIR found! May be your %s"
                   " preferences file is broken."),
            PACKAGE_NAME);
        exit(1);
    }

    magicFile = FSGetStringForName("MAGICFILE");
    if (magicFile) {
        magic_parse_file(magicFile);
        free(magicFile);
    } else
        magic_parse_file("");
}

static void
printHelp(char* progname)
{
    printf(_("Usage: %s [options]\n"), progname);
    puts(_("options:"));
    puts(_(" -p <initial path>  initial path to display"));
    puts(_(" -v                 print version number and exit"));
}

Bool FSIsFSViewerClipSet(FSViewer* fsViewer)
{
    return fsViewer->clip ? True : False;
}

void FSSetFSViewerClip(FSViewer* fsViewer, FileInfo* fileInfo)
{
    /* Need to free previous clip???? */
    /* Or need to copy fileInfo????   */
    fsViewer->clip = fileInfo;
}

FileInfo*
FSGetFSViewerClip(FSViewer* fsViewer)
{
    return fsViewer->clip;
}

void FSSetFSViewerClipAction(FSViewer* fsViewer, ClipAction action)
{
    fsViewer->clipAction = action;
}

ClipAction
FSGetFSViewerClipAction(FSViewer* fsViewer)
{
    return fsViewer->clipAction;
}

void FSSetFSViewerTransientWindow(FSViewer* fsViewer, Window window)
{
    XWMHints* hints;
    MyGNUstepWMAttributes attributes;

    // WM_DELETE should be set here...
    if ((hints = XAllocWMHints())) {
        hints->window_group = fsViewer->leader;
        hints->flags = WindowGroupHint;
        XSetWMHints(fsViewer->dpy, window, hints);
        XFree((void*)hints);
    }

    memset((void*)&attributes, 0, sizeof(MyGNUstepWMAttributes));
    attributes.window_style = (WMTitledWindowMask | WMClosableWindowMask);
    attributes.window_level = WMFloatingWindowLevel;
    attributes.extra_flags = GSFullKeyboardEventsFlag;
    attributes.flags = (GSWindowStyleAttr | GSWindowLevelAttr | GSExtraFlagsAttr);
    MyWMSetWindowAttributes(fsViewer->dpy, window, &attributes);
    WMAppAddWindow(fsViewer->wmContext, window);
}

void FSAddWindow(FSViewer* fsViewer, Window window)
{
    XWMHints* hints;
    MyGNUstepWMAttributes attributes;

    // WM_DELETE should be set here...
    if ((hints = XAllocWMHints())) {
        hints->window_group = fsViewer->leader;
        /* 	hints->flags = WindowGroupHint; */
        XSetWMHints(fsViewer->dpy, window, hints);
        XFree((void*)hints);
    }

    memset((void*)&attributes, 0, sizeof(MyGNUstepWMAttributes));
    attributes.window_style = (WMTitledWindowMask | WMClosableWindowMask | WMMiniaturizableWindowMask);
    attributes.window_level = WMNormalWindowLevel;
    attributes.extra_flags = GSFullKeyboardEventsFlag;
    attributes.flags = (GSWindowStyleAttr | GSWindowLevelAttr | GSExtraFlagsAttr);
    MyWMSetWindowAttributes(fsViewer->dpy, window, &attributes);
    WMAppAddWindow(fsViewer->wmContext, window);
}

void FSSetFSViewerConfirmWindow(FSViewer* fsViewer, Window window)
{
    XWMHints* hints;
    MyGNUstepWMAttributes attributes;

    // WM_DELETE should be set here...
    if ((hints = XAllocWMHints())) {
        hints->window_group = fsViewer->leader;
        hints->flags = WindowGroupHint;
        XSetWMHints(fsViewer->dpy, window, hints);
        XFree((void*)hints);
    }

    // This is horrible: there must be a better way?
    XStoreName(fsViewer->dpy, window, " ");
    memset((void*)&attributes, 0, sizeof(MyGNUstepWMAttributes));
    attributes.window_style = (WMTitledWindowMask);
    attributes.window_level = WMFloatingWindowLevel;
    attributes.extra_flags = GSFullKeyboardEventsFlag;
    attributes.flags = (GSWindowStyleAttr | GSWindowLevelAttr | GSExtraFlagsAttr);
    MyWMSetWindowAttributes(fsViewer->dpy, window, &attributes);
    WMAppAddWindow(fsViewer->wmContext, window);
}

WMScreen*
FSGetFSViewerScreen(FSViewer* fsViewer)
{
    return fsViewer->scr;
}

WMAppContext*
FSGetFSViewerWMContext(FSViewer* fsViewer)
{
    return fsViewer->wmContext;
}

RContext*
FSGetFSViewerRContext(FSViewer* fsViewer)
{
    return fsViewer->rcontext;
}

Display*
FSGetFSViewerDisplay(FSViewer* fsViewer)
{
    return fsViewer->dpy;
}

XContext
FSGetFSViewerXContext(FSViewer* fsViewer)
{
    return fsViewer->xContext;
}

Window
FSGetFSViewerLeader(FSViewer* fsViewer)
{
    return fsViewer->leader;
}

void FSAddViewToFSViewer(FSViewer* fsViewer, FSFileView* fView)
{
    ++fsViewer->nviews;
    fsViewer->current = fView;
}

void FSRemoveViewFromFSViewer(FSViewer* fsViewer, FSFileView* fView)
{
    --fsViewer->nviews;
}

void FSSetFSViewerFinder(FSViewer* fsViewer, FSFinder* finder)
{
    fsViewer->finder = finder;
}

FSFinder*
FSGetFSViewerFinder(FSViewer* fsViewer)
{
    return fsViewer->finder;
}

void FSSetFSViewerPath(FSViewer* fsViewer, char* path)
{
    FSSetFileViewPath(fsViewer->current, path);
}

char* FSGetFSViewerPath(FSViewer* fsViewer)
{
    FileInfo* fileInfo = NULL;
    FSFileView* fView;

    fView = FSGetFSViewerCurrentView(fsViewer);
    fileInfo = FSGetFileViewFileInfo(fView);

    return GetPathnameFromPathName(fileInfo->path, fileInfo->name);
}

FSFileView*
FSGetFSViewerCurrentView(FSViewer* fsViewer)
{
    return fsViewer->current;
}

void FSUpdateCurrentFileViewTitles()
{
    FSFileView* fView = NULL;

    /*
      Make sure we're in a window, otherwise the app
      could crash, especially if all windows are minimized,
      then one is opened and closed. When another window is
      opened, the current fileView references the one that
      was just closed, so the app crashes as that fileView
      isn't there. The current fileView isn't changed until
      a window receives the focus, therefore we check to
      see if the app is focused before unpdating the titles.
    */
    if (focusIn) {
        fView = FSGetFSViewerCurrentView(fsViewer);
        FSUpdateFileViewTitles(fView);
    }
}

int FSGetFSViewerMetaMask(FSViewer* fsViewer)
{
    return fsViewer->metaMask;
}

char* parseArgs(int argc, char** argv)
{
    char* initPath = NULL;

    if (argc > 1) {
        int i;
        for (i = 1; i < argc; i++) {
            if (strcmp(argv[i], "-v") == 0) {
                printf("FSViewer %s\n", VERSION);
                exit(0);
            } else if (strcmp(argv[i], "-p") == 0) {
                i++;
                if (i >= argc) {
                    wwarning(_("too few arguments for %s"), argv[i - 1]);
                    exit(0);
                }
                initPath = argv[i];
            } else {
                printHelp(argv[0]);
                exit(0);
            }
        }
    }

    return initPath;
}
static void
FSCreateFSViewer(FSViewer* fsViewer, char** argv, int argc)
{
    fsViewer->leader = XCreateSimpleWindow(fsViewer->dpy,
        DefaultRootWindow(fsViewer->dpy),
        10, 10, 10, 10, 0, 0, 0);

    // Set standard X hints.
    if ((fsViewer->class = XAllocClassHint())) {
        fsViewer->class->res_name = "fsviewer";
        fsViewer->class->res_class = "FSViewer";
        XSetClassHint(fsViewer->dpy, fsViewer->leader, fsViewer->class);
    }

    if ((fsViewer->hints = XAllocWMHints())) {
        fsViewer->hints->window_group = fsViewer->leader;
        fsViewer->hints->flags = WindowGroupHint;
        XSetWMHints(fsViewer->dpy, fsViewer->leader, fsViewer->hints);
        XSetCommand(fsViewer->dpy, fsViewer->leader, argv, argc);
    }

    // Context for carrying a FileView*.
    fsViewer->xContext = XUniqueContext();

    /*
     *  Create application descriptor.
     */
    WMInitializeApplication("FSViewer", &argc, argv);
    fsViewer->scr = WMCreateScreen(fsViewer->dpy,
        DefaultScreen(fsViewer->dpy));

    // Create the WorkSpace menu context.
    fsViewer->wmContext = WMAppCreateWithMain(fsViewer->dpy,
        DefaultScreen(fsViewer->dpy),
        fsViewer->leader);
    fsViewer->initPath = parseArgs(argc, argv);
    fsViewer->clip = NULL;
}

static void
FSCreateFSViewerIcons(FSViewer* fsViewer)
{
    RContextAttributes attributes;

    memset((void*)&attributes, 0, sizeof(RContextAttributes));
    attributes.flags = (RC_RenderMode | RC_ColorsPerChannel);
    attributes.render_mode = 0;
    attributes.colors_per_channel = 4;

    if (!(fsViewer->rcontext = RCreateContext(fsViewer->dpy, DefaultScreen(fsViewer->dpy),
              &attributes))) {
        /* Somethings broken here!! */
        wfatal(_("Unable to get RContext: %s %s %d\n"),
            "boo" /*RErrorString*/, __FILE__, __LINE__);
    }

    // The application icon.
    if ((fsViewer->image = RGetImageFromXPMData(fsViewer->rcontext,
             FSVIEWER_XPM))) {
        if ((RConvertImageMask(fsViewer->rcontext, fsViewer->image,
                &fsViewer->appicon, &fsViewer->appmask, 0))) {
            fsViewer->hints->icon_pixmap = fsViewer->appicon;
            fsViewer->hints->icon_mask = fsViewer->appmask;
            fsViewer->hints->flags |= (IconPixmapHint | IconMaskHint);
            XSetWMHints(fsViewer->dpy, fsViewer->leader, fsViewer->hints);
        }
    }

    // The icon used in standard dialog panels.
    if ((fsViewer->wmpixmap = WMCreatePixmapFromRImage(fsViewer->scr,
             fsViewer->image, 0))) {
        WMSetApplicationIconPixmap(fsViewer->scr, fsViewer->wmpixmap);
    }
}

static void
FSInitFSViewer(FSViewer* fsViewer)
{
    int bit;

    FSLoadFSViewerConfigurations(fsViewer);
    FSInitSystemInfo(fsViewer);
    FSInitInspector(fsViewer);
    bit = ModifierFromKey(FSGetFSViewerDisplay(fsViewer), "META");
    fsViewer->metaMask = 1 << bit;
}

int main(int argc, char** argv)
{
    FSFileView* fView;
    focusIn = False;

    setlocale(LC_ALL, "");

#ifdef ENABLE_NLS
    if (getenv("NLSPATH"))
        bindtextdomain("FSViewer", getenv("NLSPATH"));
    else
        bindtextdomain("FSViewer", LOCALEDIR);
    textdomain("FSViewer");

    if (!XSupportsLocale()) {
        wwarning(_("X server does not support locale"));
    }
    if (XSetLocaleModifiers("") == NULL) {
        wwarning(_("cannot set locale modifiers"));
    }
#endif

    /* wmessage("LOCALEDIR= %s", LOCALEDIR);*/

    if (!(fsViewer = (FSViewer*)malloc(sizeof(FSViewer)))) {
        wfatal(_("Unable to allocate memory for %s %s %d\n"),
            argv[0], __FILE__, __LINE__);
    }
    memset((void*)fsViewer, 0, sizeof(FSViewer));

    fsViewer->dpy = XOpenDisplay(NULL);
    if (!fsViewer->dpy) {
        wfatal(_("Unable to Open Display!"));
        exit(0);
    }
    FSCreateFSViewer(fsViewer, argv, argc);
    FSCreateFSViewerIcons(fsViewer);
    FSInitFSViewer(fsViewer);
    FSCreateMenu(fsViewer);

    /*
       No path specified at start up, use home dir.
    */
    if (fsViewer->initPath == NULL)
        fsViewer->initPath = FSGetHomeDir();
    /*
       Unable to get home dir, must use `/`
    */
    if (fsViewer->initPath == NULL)
        fsViewer->initPath = wstrdup("/");

    if (!(fView = FSCreateFileView(fsViewer, fsViewer->initPath, True))) {
        wfatal(_("Unable to create a FileView instance %s %s %d"),
            argv[0], __FILE__, __LINE__);
    }
    fsViewer->current = fView;

    while (1) {
        XEvent event;
        FSFileView* fView;

        WMNextEvent(fsViewer->dpy, &event);
        WMHandleEvent(&event);
        WMProcessEvent(fsViewer->wmContext, &event);

        switch (event.type) {
        case FocusIn:
            if (!(XFindContext(fsViewer->dpy, event.xfocus.window,
                    fsViewer->xContext, (XPointer*)&fView))) {
                /* 		    if((FSFinder *)fView == fsViewer->finder) */
                /* 			break; */
                fsViewer->current = fView;
                FSUpdateFileViewShelf(fView);
            }
            focusIn = True;
            break;
        case FocusOut:
            focusIn = False;
            break;
        case ClientMessage:
            // printf("arrgggg!\n");
            break;
        }
    }
}
